src/decoder/mod.rs

```diff
@@ -611,63 +734,33 @@ where
+            if self.settings.gapless
+                && self.settings.is_seekable
+                && inner.try_seek(Duration::ZERO).is_ok()
+            {
+                return inner.next();
+            }
+
+            // Fall back to recreation with cached metadata to avoid expensive scanning
             let decoder = self.inner.take()?;
-            let (new_decoder, sample) = match decoder {
-                #[cfg(all(feature = "hound", not(feature = "symphonia-wav")))]
-                DecoderImpl::Wav(source) => {
-                    let mut reader = source.into_inner();
-                    reader.seek(SeekFrom::Start(0)).ok()?;
-                    let mut source = wav::WavDecoder::new(reader).ok()?;
-                    let sample = source.next();
-                    (DecoderImpl::Wav(source), sample)
-                }
-                #[cfg(all(feature = "lewton", not(feature = "symphonia-vorbis")))]
-                DecoderImpl::Vorbis(source) => {
-                    use lewton::inside_ogg::OggStreamReader;
-                    let mut reader = source.into_inner().into_inner();
-                    reader.seek_bytes(SeekFrom::Start(0)).ok()?;
-                    let mut source = vorbis::VorbisDecoder::from_stream_reader(
-                        OggStreamReader::from_ogg_reader(reader).ok()?,
-                    );
-                    let sample = source.next();
-                    (DecoderImpl::Vorbis(source), sample)
-                }
-                #[cfg(all(feature = "claxon", not(feature = "symphonia-flac")))]
-                DecoderImpl::Flac(source) => {
-                    let mut reader = source.into_inner();
-                    reader.seek(SeekFrom::Start(0)).ok()?;
-                    let mut source = flac::FlacDecoder::new(reader).ok()?;
-                    let sample = source.next();
-                    (DecoderImpl::Flac(source), sample)
-                }
-                #[cfg(all(feature = "minimp3", not(feature = "symphonia-mp3")))]
-                DecoderImpl::Mp3(source) => {
-                    let mut reader = source.into_inner();
-                    reader.seek(SeekFrom::Start(0)).ok()?;
-                    let mut source = mp3::Mp3Decoder::new(reader).ok()?;
-                    let sample = source.next();
-                    (DecoderImpl::Mp3(source), sample)
-                }
-                #[cfg(feature = "symphonia")]
-                DecoderImpl::Symphonia(source, PhantomData) => {
-                    let mut reader = source.into_inner();
-                    reader.seek(SeekFrom::Start(0)).ok()?;
-                    let mut source =
-                        symphonia::SymphoniaDecoder::new(reader, &self.settings).ok()?;
-                    let sample = source.next();
-                    (DecoderImpl::Symphonia(source, PhantomData), sample)
-                }
-            };
+            let (new_decoder, sample) = self.recreate_decoder_with_cache(decoder)?;
```

Comment:
Very nice, much more readable then the previous code. I'm wondering if it makes sense to lift looped decoder to its own module? But we can always do that later.